src/mlpack/core/cereal/is_loading.hpp

```diff
@@ -0,0 +1,56 @@
+/**
+ * @file core/cereal/is_loading.hpp
+ * @author Ryan Curtin
+ *
+ * Implementation of is_loading function.
+ *
+ * This implementation provides backward compatibilty with older
+ * version of cereal that does not have Archive::is_loading struct. 
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_CORE_CEREAL_IS_LOADING_HPP
+#define MLPACK_CORE_CEREAL_IS_LOADING_HPP
+
+#include <cereal/archives/binary.hpp>
+#include <cereal/archives/portable_binary.hpp>
+#include <cereal/archives/xml.hpp>
+#include <cereal/archives/json.hpp>
+
+namespace cereal {
+
+template<typename Archive>
+struct is_cereal_archive
+{
+  // Archive::is_loading is not implemented yet, so we can use std::is_same<>
+    // to check if it is a loading archive.
+    constexpr static bool value = std::is_same<Archive, cereal::BinaryInputArchive>::value ||
+#if (BINDING_TYPE != BINDING_TYPE_R)
+std::is_same<Archive, cereal::JSONInputArchive>::value ||
+#endif
```

Comment:
@shrit here's the problem.  I don't understand why this is needed; it looks like part of the patch for R bindings that you applied.  @Yashwants19 why was this needed for R?  I wonder if `BINDING_TYPE` is not defined for the tests we are building, and thus the `#if` doesn't work.  (If we do need it, I guess maybe we could make the check as `#if !defined(BINDING_TYPE) || (BINDING_TYPE != BINDING_TYPE_R)` or something like this.)

Reply:
If I didn't use `#define CEREAL_ARCHIVES_JSON_HPP_` in `inst/include/mlpack.h` I got these error, hence to resolve this error I have to use this definition in `inst/include/mlpack.h` and `src/rcpp_mlpack.h`and bypass all the `JSON` calls, I am really sorry, I didn't discussion about the solution, may be you might some better solution here:

```
In file included from /usr/share/R/include/R.h:91,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/Rcpp/r/headers.h:71,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/RcppCommon.h:29,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/Rcpp.h:27,
                 from ../inst/include/mlpack.h:15,
                 from RcppExports.cpp:4:
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:71:66: error: expected ‘)’ before ‘newSize’
   71 |     void* Realloc(void* originalPtr, size_t originalSize, size_t newSize) {
      |                                                                  ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:25: note: in definition of macro ‘Realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                         ^
/usr/share/R/include/R_ext/RS.h:73:24: note: to match this ‘(’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:71:11: note: in expansion of macro ‘Realloc’
   71 |     void* Realloc(void* originalPtr, size_t originalSize, size_t newSize) {
      |           ^~~~~~~
/usr/share/R/include/R_ext/RS.h:74:37: error: expected identifier before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                                     ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:79:17: note: in expansion of macro ‘Free’
   79 |     static void Free(void *ptr) { std::free(ptr); }
      |                 ^~~~
/usr/share/R/include/R_ext/RS.h:74:37: error: ‘parameter’ declared as function returning a function
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                                     ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:79:17: note: in expansion of macro ‘Free’
   79 |     static void Free(void *ptr) { std::free(ptr); }
      |                 ^~~~
/usr/share/R/include/R_ext/RS.h:74:50: error: expected ‘)’ before ‘,’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ~                         ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:79:17: note: in expansion of macro ‘Free’
   79 |     static void Free(void *ptr) { std::free(ptr); }
      |                 ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:191:66: error: expected ‘)’ before ‘newSize’
  191 |     void* Realloc(void* originalPtr, size_t originalSize, size_t newSize) {
      |                                                                  ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:25: note: in definition of macro ‘Realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                         ^
/usr/share/R/include/R_ext/RS.h:73:24: note: to match this ‘(’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:191:11: note: in expansion of macro ‘Realloc’
  191 |     void* Realloc(void* originalPtr, size_t originalSize, size_t newSize) {
      |           ^~~~~~~
/usr/share/R/include/R_ext/RS.h:74:37: error: expected identifier before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                                     ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:225:17: note: in expansion of macro ‘Free’
  225 |     static void Free(void *ptr) { (void)ptr; } // Do nothing
      |                 ^~~~
/usr/share/R/include/R_ext/RS.h:74:37: error: ‘parameter’ declared as function returning a function
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                                     ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:225:17: note: in expansion of macro ‘Free’
  225 |     static void Free(void *ptr) { (void)ptr; } // Do nothing
      |                 ^~~~
/usr/share/R/include/R_ext/RS.h:74:50: error: expected ‘)’ before ‘,’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ~                         ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:225:17: note: in expansion of macro ‘Free’
  225 |     static void Free(void *ptr) { (void)ptr; } // Do nothing
      |                 ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h: In member function ‘void rapidjson::MemoryPoolAllocator<BaseAllocator>::Clear()’:
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:148:29: note: in expansion of macro ‘Free’
  148 |             baseAllocator_->Free(chunkHead_);
      |                             ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h: In member function ‘void rapidjson::internal::Stack<Allocator>::ShrinkToFit()’:
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:103:24: note: in expansion of macro ‘Free’
  103 |             Allocator::Free(stack_);
      |                        ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h: In member function ‘void rapidjson::internal::Stack<Allocator>::Resize(std::size_t)’:
/usr/share/R/include/R_ext/RS.h:73:24: error: expected unqualified-id before ‘(’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:201:49: note: in expansion of macro ‘Realloc’
  201 |         stack_ = static_cast<char*>(allocator_->Realloc(stack_, GetCapacity(), newCapacity));
      |                                                 ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:28: error: expected primary-expression before ‘)’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                            ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:201:49: note: in expansion of macro ‘Realloc’
  201 |         stack_ = static_cast<char*>(allocator_->Realloc(stack_, GetCapacity(), newCapacity));
      |                                                 ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:30: error: expected ‘)’ before ‘R_chk_realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                              ^~~~~~~~~~~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:201:49: note: in expansion of macro ‘Realloc’
  201 |         stack_ = static_cast<char*>(allocator_->Realloc(stack_, GetCapacity(), newCapacity));
      |                                                 ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:30: error: expected ‘;’ before ‘R_chk_realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                              ^~~~~~~~~~~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:201:49: note: in expansion of macro ‘Realloc’
  201 |         stack_ = static_cast<char*>(allocator_->Realloc(stack_, GetCapacity(), newCapacity));
      |                                                 ^~~~~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h: In member function ‘void rapidjson::internal::Stack<Allocator>::Destroy()’:
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/stack.h:207:20: note: in expansion of macro ‘Free’
  207 |         Allocator::Free(stack_);
      |                    ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h: In destructor ‘rapidjson::GenericValue<Encoding, Allocator>::~GenericValue()’:
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:727:32: note: in expansion of macro ‘Free’
  727 |                     Allocator::Free(e);
      |                                ^~~~
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:734:28: note: in expansion of macro ‘Free’
  734 |                 Allocator::Free(GetMembersPointer());
      |                            ^~~~
/usr/share/R/include/R_ext/RS.h:74:24: error: expected unqualified-id before ‘(’ token
   74 | #define Free(p)        (R_chk_free( (void *)(p) ), (p) = NULL)
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:738:28: note: in expansion of macro ‘Free’
  738 |                 Allocator::Free(const_cast<Ch*>(GetStringPointer()));
      |                            ^~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h: In member function ‘rapidjson::GenericValue<Encoding, Allocator>& rapidjson::GenericValue<Encoding, Allocator>::AddMember(rapidjson::GenericValue<Encoding, Allocator>&, rapidjson::GenericValue<Encoding, Allocator>&, Allocator&)’:
/usr/share/R/include/R_ext/RS.h:73:24: error: expected unqualified-id before ‘(’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1199:71: note: in expansion of macro ‘Realloc’
 1199 |                 SetMembersPointer(reinterpret_cast<Member*>(allocator.Realloc(GetMembersPointer(), oldCapacity * sizeof(Member), o.capacity * sizeof(Member))));
      |                                                                       ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:28: error: expected primary-expression before ‘)’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                            ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1199:71: note: in expansion of macro ‘Realloc’
 1199 |                 SetMembersPointer(reinterpret_cast<Member*>(allocator.Realloc(GetMembersPointer(), oldCapacity * sizeof(Member), o.capacity * sizeof(Member))));
      |                                                                       ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:30: error: expected ‘)’ before ‘R_chk_realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                              ^~~~~~~~~~~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1199:71: note: in expansion of macro ‘Realloc’
 1199 |                 SetMembersPointer(reinterpret_cast<Member*>(allocator.Realloc(GetMembersPointer(), oldCapacity * sizeof(Member), o.capacity * sizeof(Member))));
      |                                                                       ^~~~~~~
In file included from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/archives/json.hpp:56,
                 from ./mlpack/prereqs.hpp:80,
                 from ./mlpack/core/data/extension.hpp:16,
                 from ./mlpack/core/data/detect_file_type.cpp:15,
                 from ../inst/include/mlpack.h:46,
                 from RcppExports.cpp:4:
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1199:159: error: expected ‘;’ before ‘)’ token
 1199 |                 SetMembersPointer(reinterpret_cast<Member*>(allocator.Realloc(GetMembersPointer(), oldCapacity * sizeof(Member), o.capacity * sizeof(Member))));
      |                                                                                                                                                               ^
      |                                                                                                                                                               ;
In file included from /usr/share/R/include/R.h:91,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/Rcpp/r/headers.h:71,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/RcppCommon.h:29,
                 from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcpp/include/Rcpp.h:27,
                 from ../inst/include/mlpack.h:15,
                 from RcppExports.cpp:4:
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h: In member function ‘rapidjson::GenericValue<Encoding, Allocator>& rapidjson::GenericValue<Encoding, Allocator>::Reserve(rapidjson::SizeType, Allocator&)’:
/usr/share/R/include/R_ext/RS.h:73:24: error: expected unqualified-id before ‘(’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1525:74: note: in expansion of macro ‘Realloc’
 1525 |             SetElementsPointer(reinterpret_cast<GenericValue*>(allocator.Realloc(GetElementsPointer(), data_.a.capacity * sizeof(GenericValue), newCapacity * sizeof(GenericValue))));
      |                                                                          ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:28: error: expected primary-expression before ‘)’ token
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                            ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1525:74: note: in expansion of macro ‘Realloc’
 1525 |             SetElementsPointer(reinterpret_cast<GenericValue*>(allocator.Realloc(GetElementsPointer(), data_.a.capacity * sizeof(GenericValue), newCapacity * sizeof(GenericValue))));
      |                                                                          ^~~~~~~
/usr/share/R/include/R_ext/RS.h:73:30: error: expected ‘)’ before ‘R_chk_realloc’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                              ^~~~~~~~~~~~~
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1525:74: note: in expansion of macro ‘Realloc’
 1525 |             SetElementsPointer(reinterpret_cast<GenericValue*>(allocator.Realloc(GetElementsPointer(), data_.a.capacity * sizeof(GenericValue), newCapacity * sizeof(GenericValue))));
      |                                                                          ^~~~~~~
In file included from /home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/archives/json.hpp:56,
                 from ./mlpack/prereqs.hpp:80,
                 from ./mlpack/core/data/extension.hpp:16,
                 from ./mlpack/core/data/detect_file_type.cpp:15,
                 from ../inst/include/mlpack.h:46,
                 from RcppExports.cpp:4:
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/document.h:1525:181: error: expected ‘;’ before ‘)’ token
 1525 |             SetElementsPointer(reinterpret_cast<GenericValue*>(allocator.Realloc(GetElementsPointer(), data_.a.capacity * sizeof(GenericValue), newCapacity * sizeof(GenericValue))));
      |                                                                                                                                                                                     ^
      |                                                                                                                          
```

Reply:
@rcurtin Thanks, exactly I have arrived at the same results yesterday night. I have removed the binding declaration, I am working on the `CoverTreeTest`. :+1:

Reply:
@Yashwants19 looks like a macro collision:

```
/usr/share/R/include/R_ext/RS.h:73:24: note: to match this ‘(’
   73 | #define Realloc(p,n,t) (t *) R_chk_realloc( (void *)(p), (R_SIZE_T)((n) * sizeof(t)) )
      |                        ^
/home/yashwant/R/x86_64-pc-linux-gnu-library/4.0/Rcereal/include/cereal/external/rapidjson/internal/../allocators.h:71:11: note: in expansion of macro ‘Realloc’
   71 |     void* Realloc(void* originalPtr, size_t originalSize, size_t newSize) {
```

Maybe before including the cereal headers, you want to, e.g., `#undef Realloc` and then redefine it after inclusion is done?

Reply:
`#if !defined(BINDING_TYPE) || (BINDING_TYPE != BINDING_TYPE_R)`
I think so this also looks like a nice solution.
Or should we use `#undef  Realloc` and `#undef Free` for fixing this?

Reply:
Ahh, `Free` overlaps too.  Personally I think the `#undef` solution is better, because then users will still be able to serialize to JSON.  The `#if` solution will disable JSON serialization.  Maybe give it a shot and see if it works?  I assume after we include cereal, we'll have to then redefine `Realloc` and `Free`.

Reply:
I have try to apply `#undef` solution in this patch (https://gist.github.com/Yashwants19/6cdadffb26549c07c545033ca17188a3).
Let me know what you think?